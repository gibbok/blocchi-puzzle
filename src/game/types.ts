export enum DirectionEnum {
  N = 'N',
  E = 'E',
  S = 'S',
  W = 'W',
}

export const NO = DirectionEnum.N;
export const ES = DirectionEnum.E;
export const SO = DirectionEnum.S;
export const WE = DirectionEnum.W;

export enum TetroEnum {
  Z = 'Z',
  S = 'S',
  J = 'J',
  T = 'T',
  I = 'I',
  L = 'L',
  O = 'O',
}

export enum NoTetroEnum {
  NoTetro,
}
export const NoTetro = NoTetroEnum.NoTetro;

export type Tile = NoTetroEnum.NoTetro | TetroEnum;

export type TetroPieces = Record<TetroEnum, Record<DirectionEnum, Tetro>>;

export type TetroRow = readonly Tile[];

// The tetromino, a geometric shape composed of four squares/tiles
export type Tetro = readonly TetroRow[];

export type TetroDef = Readonly<{
  type: TetroEnum;
  direction: DirectionEnum;
  x: number;
  y: number;
}>;

export const Z = TetroEnum.Z;
export const S = TetroEnum.S;
export const J = TetroEnum.J;
export const T = TetroEnum.T;
export const I = TetroEnum.I;
export const L = TetroEnum.L;
export const O = TetroEnum.O;

export type Position = number;

export type BoardRow = readonly Tile[];

export type BoardMutable = Tile[][];

// This is the space on which the game is played.
// Tetriminos fall from the top of the Board, and then players turn and move them into their desired places at the bottom.
export type Board = readonly BoardRow[];

export enum ScreenEnum {
  Intro = 'Intro',
  Game = 'Game',
  Over = 'Over',
}

export type InternalState = Readonly<{
  board: Board;
  score: number; // some point based on lines done
  level: number; // based on score, every 1000 score you go to next level, speed is faster
  lines: number; // number of lines done
  currentTetro: TetroDef;
  nextTetro: TetroDef;
  isPlay: boolean;
  isGameOver: boolean;
  screen: ScreenEnum;
}>;

export type PublicState = Omit<InternalState, 'currentTetro' | 'isPlay' | 'isGameOver'>;

export enum KeyEnum {
  Esc = 27,
  Space = 32,
  Left = 37,
  Up = 38,
  Right = 39,
  Down = 40,
}

export type CallBack = () => void;

export type Action = Readonly<{
  type: string;
}>;

export type TransitionState = 'entering' | 'entered' | 'exiting' | 'exited';
export interface TransitionStyles {
  entering?: React.CSSProperties;
  entered?: React.CSSProperties;
  exiting?: React.CSSProperties;
  exited?: React.CSSProperties;
}
