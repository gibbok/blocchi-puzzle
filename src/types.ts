export enum DirectionEnum {
  N = 'N',
  E = 'E',
  S = 'S',
  W = 'W'
}

export const NO = DirectionEnum.N;
export const ES = DirectionEnum.E;
export const SO = DirectionEnum.S;
export const WE = DirectionEnum.W;

export enum TetroEnum {
  Z = 'Z',
  S = 'S',
  J = 'J',
  T = 'T',
  I = 'I',
  O = 'O'
}

export type NoTetro = 0;
export type Block = NoTetro | TetroEnum;

export type TetroPieces = Record<TetroEnum, Record<DirectionEnum, Tetro>>;

export type TetroRow = readonly Block[];

export type Tetro = readonly (TetroRow)[];

export type TetroDef = Readonly<{
  type: TetroEnum;
  direction: DirectionEnum;
  x: number;
  y: number;
}>;

export const Z = TetroEnum.Z;
export const S = TetroEnum.S;
export const J = TetroEnum.J;
export const T = TetroEnum.T;
export const I = TetroEnum.I;
export const O = TetroEnum.O;

export type Position = number;

export type BoardRow = readonly Block[];

export type Board = readonly (BoardRow)[];

export type InternalState = Readonly<{
  board: Board;
  score: number;
  level: number;
  lines: number;
  currentTetro: TetroDef;
  nextTetro: TetroDef;
  isPlay: boolean;
}>;

export type PubicState = Omit<InternalState, 'currentTetro' | 'isPlay'>;

export enum KeyEnum {
  Esc = 27,
  Space = 32,
  Left = 37,
  Up = 38,
  Right = 39,
  Down = 40
}

export type CallBack = () => void;

export type Action = Readonly<{
  type: string;
}>;

export enum ActionEnum {
  MoveDown = 'MoveDown',
  MoveRight = 'MoveRight',
  MoveLeft = 'MoveLeft',
  MoveUp = 'MoveUp'
}
