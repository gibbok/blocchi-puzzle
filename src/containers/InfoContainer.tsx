import { connect } from 'react-redux';
import { mkPublicState } from '../store/reducer';
import { Info } from '../components/Info';
import { InternalState } from '~game/types';

const mapStateToProps = (
  prevState: InternalState
): { score: number; level: number; lines: number } => {
  const { score, level, lines } = mkPublicState(prevState);
  return { score, level, lines };
};

export const InfoContainer = connect(mapStateToProps)(Info);
