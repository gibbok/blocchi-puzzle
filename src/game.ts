export const calcScore = (rowCompleted: number) => rowCompleted * 100;

export const calcLevel = (score: number) => Math.ceil((score + 1) / 1000);
