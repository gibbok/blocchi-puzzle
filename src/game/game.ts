/**
 * Calculate the game score.
 * @param rowCompleted Number of rows completed
 */
export const calcScore = (rowCompleted: number): number => rowCompleted * 100;

/**
 * Calculate the game level.
 * @param score Game score
 */
export const calcLevel = (score: number): number => Math.ceil((score + 1) / 500);
