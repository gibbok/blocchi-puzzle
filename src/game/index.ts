export * from './board';
export * from './game';
export * from './interaction';
export * from './pieces';
export * from './tetromino';
export * from './types';
