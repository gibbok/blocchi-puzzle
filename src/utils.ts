import { Tetro, Board } from './types';

export const logger = (x: Tetro | Board) => {
  console.log(x.reduce((acc, v) => `${acc} ${v.concat()} \n`, ''));
};
