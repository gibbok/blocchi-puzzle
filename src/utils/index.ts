export * from './data.support.test';
export * from './logger';
