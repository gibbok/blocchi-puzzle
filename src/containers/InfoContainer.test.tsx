import { mapStateToProps, InfoContainer } from './InfoContainer';
import { TetroEnum } from '../game/types';
import { mkInitialState } from '../store';
import renderer from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import { mockStore } from '../utils';

// http://rahulgaba.com/front-end/2018/10/19/unit-testing-redux-containers-the-better-way-using-jest.html
// https://www.robinwieruch.de/react-connected-component-test
describe('<InfoContainer />', () => {
  it('should mapStateToProps correctly', () => {
    const input = mkInitialState(TetroEnum.I, TetroEnum.J);
    const output = mapStateToProps(input);
    const result = { score: 0, level: 1, lines: 0 };

    expect(output).toMatchObject(result);
  });

  it('should render correctly', () => {
    const store = mockStore();
    const withProvider = (
      <Provider store={store}>
        <InfoContainer />
      </Provider>
    );
    const component = renderer.create(withProvider);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
