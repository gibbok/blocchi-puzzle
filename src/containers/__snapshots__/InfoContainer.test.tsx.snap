// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoContainer /> should render correctly 1`] = `
<div
  className="sc-AxjAm jdsAYC"
>
  <div
    className="sc-AxirZ gVRNow"
  >
    <div
      className="sc-AxiKw eqhIAV"
    >
      Score
    </div>
    <div
      className="sc-AxiKw eqhIAV"
      data-test="score"
    >
      0
    </div>
  </div>
  <div
    className="sc-AxhCb jNrBnE"
  />
  <div
    className="sc-AxirZ gVRNow"
  >
    <div
      className="sc-AxiKw eqhIAV"
    >
      Level
    </div>
    <div
      className="sc-AxiKw eqhIAV"
      data-test="level"
    >
      1
    </div>
  </div>
  <div
    className="sc-AxhCb jNrBnE"
  />
  <div
    className="sc-AxirZ gVRNow"
  >
    <div
      className="sc-AxiKw eqhIAV"
    >
      Lines
    </div>
    <div
      className="sc-AxiKw eqhIAV"
      data-test="lines"
    >
      0
    </div>
  </div>
</div>
`;
